src/workspace/mod.rs

```diff
@@ -0,0 +1,1823 @@
+        root_config.to_import_map_specifier()
+      })
+      .unwrap_or(Ok(None))
+  }
+
+  pub fn to_lock_config(&self) -> Result<Option<LockConfig>, AnyError> {
+    self
+      .with_root_config_only(|root_config| root_config.to_lock_config())
+      .unwrap_or(Ok(None))
+  }
+
+  pub fn to_maybe_imports(&self) -> Result<Vec<(Url, Vec<String>)>, AnyError> {
+    self
+      .with_root_config_only(|root_config| root_config.to_maybe_imports())
+      .unwrap_or(Ok(Vec::new()))
+  }
+
+  pub fn to_maybe_jsx_import_source_config(
+    &self,
+  ) -> Result<Option<JsxImportSourceConfig>, AnyError> {
+    self
+      .with_root_config_only(|root_config| {
+        root_config.to_maybe_jsx_import_source_config()
+      })
+      .unwrap_or(Ok(None))
+  }
+
+  pub fn resolve_ctxs_from_patterns(
+    &self,
+    patterns: &FilePatterns,
+  ) -> Vec<WorkspaceMemberContext> {
+    let context = self.resolve_start_ctx();
+    let mut ctxs = Vec::with_capacity(self.config_folders.len());
+    ctxs.push(context);
+
+    // sub configs
+    for (dir_url, folder) in self.config_folders.iter() {
+      let config = match &folder.deno_json {
+        Some(c) => c,
+        None => continue,
+      };
+      if *dir_url != self.start_dir
+        && dir_url.as_str().starts_with(self.start_dir.as_str())
+        && patterns.matches_path(&config.dir_path(), PathKind::Directory)
+      {
+        let member_ctx = self.resolve_member_ctx(dir_url);
+        ctxs.push(member_ctx);
+      }
+    }
+
+    ctxs
+  }
+
+  pub fn resolve_config_excludes(&self) -> Result<PathOrPatternSet, AnyError> {
+    // have the root excludes at the front because they're lower priority
+    let mut excludes = match &self.root_folder().1.deno_json {
+      Some(c) => c.to_files_config()?.exclude.into_path_or_patterns(),
+      None => Default::default(),
+    };
+    for (dir_url, folder) in self.config_folders.iter() {
+      let Some(deno_json) = folder.deno_json.as_ref() else {
+        continue;
+      };
+      if dir_url == &self.root_dir {
+        continue;
+      }
+      excludes
+        .extend(deno_json.to_files_config()?.exclude.into_path_or_patterns());
+    }
+    Ok(PathOrPatternSet::new(excludes))
+  }
+
+  pub fn unstable_features(&self) -> &[String] {
+    self
+      .root_folder()
+      .1
+      .deno_json
+      .as_ref()
+      .map(|c| (&c.json.unstable) as &[String])
+      .unwrap_or(&[])
+  }
+
+  pub fn has_unstable(&self, name: &str) -> bool {
+    self
+      .root_folder()
+      .1
+      .deno_json
+      .as_ref()
+      .map(|c| c.has_unstable(name))
+      .unwrap_or(false)
+  }
+
+  fn with_root_config_only<R>(
+    &self,
+    with_root: impl Fn(&ConfigFile) -> R,
+  ) -> Option<R> {
+    let configs = self.config_folders.get(&self.root_dir).unwrap();
+    configs.deno_json.as_ref().map(|c| with_root(c))
+  }
+}
+
+pub enum TaskOrScript<'a> {
+  Task(&'a IndexMap<String, Task>, &'a str),
+  Script(&'a IndexMap<String, String>, &'a str),
+}
```

Comment:
Can you add a docstring describing the difference between the two? Are these `deno.json` tasks and `package.json` scripts?

Reply:
As mentioned in the call, it would be nice if the tasks defined in the workspace root had `(workspace)` annotation the same way we add `(from package.json)`. Not sure if that info needs to be stored here, but leaving for posterity

Reply:
`WorkspaceTasksConfig` has if it's in the root or member.

Reply:
`WorkspaceTasksConfig` has if it's in the root or member.